ocaml/testsuite/tests/typing-layouts/allow-illegal-crossing.ml

```diff
@@ -0,0 +1,654 @@
+type t = Foo of string | Bar of int
+val x : t = Foo "hello world"
+|}]
+
+(********************************************************)
+(* Test 3: types cannot cross other axes when annotated *)
+
+type a : value mod local
+type b : value mod global = private a
+[%%expect {|
+type a : value mod local
+Line 2, characters 0-37:
+2 | type b : value mod global = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a is value, because
+         of the definition of a at line 1, characters 0-24.
+       But the layout of type a must be a sublayout of value, because
+         of the definition of b at line 2, characters 0-37.
+|}]
+
+type a : value mod once
+type b : value mod many = private a
+[%%expect {|
+type a : value mod once
+Line 2, characters 0-35:
+2 | type b : value mod many = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a is value, because
+         of the definition of a at line 1, characters 0-23.
+       But the layout of type a must be a sublayout of value, because
+         of the definition of b at line 2, characters 0-35.
+|}]
+
+type a : value mod shared
+type b : value mod unique = private a
+[%%expect {|
+type a : value mod shared
+Line 2, characters 0-37:
+2 | type b : value mod unique = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a is value, because
+         of the definition of a at line 1, characters 0-25.
+       But the layout of type a must be a sublayout of value, because
+         of the definition of b at line 2, characters 0-37.
+|}]
+
+type a : value
+type b : immediate = private a
+[%%expect {|
+type a : value
+Line 2, characters 0-30:
+2 | type b : immediate = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a is value, because
+         of the definition of a at line 1, characters 0-14.
+       But the layout of type a must be a sublayout of immediate, because
+         of the definition of b at line 2, characters 0-30.
+|}]
+
+module _ = struct
+  type a : value mod local
+  type b : value mod global = private a
+end
+[%%expect {|
+Line 3, characters 2-39:
+3 |   type b : value mod global = private a
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a/2 is value, because
+         of the definition of a at line 2, characters 2-26.
+       But the layout of type a/2 must be a sublayout of value, because
+         of the definition of b at line 3, characters 2-39.
+|}]
+
+module A : sig
+  type t
+end = struct
+  type t : value mod many = private string
+end
+[%%expect {|
+Line 4, characters 2-42:
+4 |   type t : value mod many = private string
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type string is value, because
+         it is the primitive value type string.
+       But the layout of type string must be a sublayout of value, because
+         of the definition of t at line 4, characters 2-42.
+|}]
+
+type 'a t : value mod unique = private 'a
```

Comment:
I'm surprised this is rejected. I would just expect to infer a `value mod unique` kind for `'a`. I think this deserves investigation.

Reply:
Good point. This bug appears to have already been present in main, before my changes:
```ocaml
type 'a t : word = 'a
[%%expect {|
Line 1, characters 0-21:
1 | type 'a t : word = 'a
    ^^^^^^^^^^^^^^^^^^^^^
Error: The layout of type 'a is value, because
         of the definition of t at line 1, characters 0-21.
       But the layout of type 'a must be a sublayout of word, because
         of the definition of t at line 1, characters 0-21.
|}]
```
So I agree this needs to get fixed, but I believe it is outside the scope of this PR. For now, I'll remove this test.

Reply:
I added some tests demonstrating this bug to `jkinds.ml`